src/TinyCiv.Server/Handlers/UnitMoveHandler.cs

```diff
@@ -0,0 +1,114 @@
+﻿using Microsoft.AspNetCore.SignalR;
+using TinyCiv.Server.Core.Extensions;
+using TinyCiv.Server.Core.Services;
+using TinyCiv.Shared;
+using TinyCiv.Shared.Events.Client;
+using TinyCiv.Shared.Events.Server;
+using TinyCiv.Shared.Game;
+
+namespace TinyCiv.Server.Handlers
+{
+    public class UnitMoveHandler : ClientHandler<MoveUnitClientEvent>
+    {
+        private readonly IMapService _mapService;
+
+        private readonly object _movementLocker = new();
+
+        public UnitMoveHandler(IMapService mapService)
+        {
+            _mapService = mapService;
+        }
+
+        protected override async Task OnHandleAsync(IClientProxy caller, IClientProxy all, MoveUnitClientEvent @event)
+        {
+            var unitMovementTuple = (@event.UnitId, new ServerPosition { X = @event.X, Y = @event.Y });
+
+            lock (_movementLocker)
+            {
+                // Yes, it is being checked 2 times :| Just put the validation into another function?
+                if (IgnoreWhen(@event))
+                {
+                    return;
+                }
+
+                _mapService.MovingUnits.Add(unitMovementTuple);
+            }
+
+            await caller
+                .SendEventAsync(Constants.Server.SendUnitStatusUpdate, new UnitStatusUpdateServerEvent(@event.UnitId, true))
+                .ConfigureAwait(false);
+
+            var unit = _mapService.GetUnit(@event.UnitId);
+
+            if (unit == null)
+            {
+                return;
+            }
+
+            _ = Task.Run(async () =>
```

Comment:
This does not need locking?

For this for example?
```_mapService.MovingUnits.Remove(unitMovementTuple)```

Reply:
I don't remember, but isn't there some structures that automatically handle locking, as in it would modify the collection only on one thread at the time

My mistake, should've probably researched that first